website/blog/2023-05-28-databend-weekly-95.md

```diff
@@ -0,0 +1,160 @@
+
+If you are interested in learning more, please check out the resources listed below:
+
+- [PR | feat: support IEJoin](https://github.com/datafuselabs/databend/pull/11412)
+- [Paper | Lightning Fast and Space Efficient Inequality Joins](https://wayang.apache.org/assets/pdf/paper/iejoin.pdf)
+
+## Code Corner
+
+Discover some fascinating code snippets or projects that showcase our work or learning journey.
+
+### Python Binding for Databend Rust Driver
+
+PyO3 provides bindings between Rust and Python, along with a ready-to-use scaffolding tool called Maturin. With Maturin, it's easy to create Python extension modules based on Rust. 
+
+This allows for code reorganization where performance-critical parts can be rewritten in Rust while the rest of the code can still use original Python code.
+
+Based on PyO3, we can focus more conveniently on the Rust implementation logic itself without paying too much attention to FFI (Foreign Function Interface) and conversion details. This makes it possible to convert Rust into a Python module at a lower cost, and only maintain one set of code, greatly reducing maintenance costs. Databend is implementing Python bindings for Rust drivers, with the ultimate goal of replacing the current databend-py. 
+
+Here are some code snippets that may help you understand how to use PyO3.
+
+In `asyncio.rs`, a Struct called `AsyncDatabendDriver` can be defined as a Python class, and the Python module can be defined as `databend-driver`. 
+
+```rust
+/// `AsyncDatabendDriver` is the entry for all public async API
+#[pyclass(module = "databend_driver")]
+pub struct AsyncDatabendDriver(Connector);
+```
+
+The corresponding methods for `AsyncDatabendDriver` are implemented next, with underlying calls made to the methods implemented in Rust's Trait. For example, in this code snippet, we see that `exec` is used:
+
+```rust
+#[pymethods]
+impl AsyncDatabendDriver {
+    #[new]
+    #[pyo3(signature = (dsn))]
+    pub fn new(dsn: &str) -> PyResult<Self> {
+        Ok(AsyncDatabendDriver(build_connector(dsn)?))
+    }
+
+    /// exec
+    pub fn exec<'p>(&'p self, py: Python<'p>, sql: String) -> PyResult<&'p PyAny> {
+        let this = self.0.clone();
+        future_into_py(py, async move {
+            let res = this.connector.exec(&sql).await.unwrap();
+            Ok(res)
+        })
+    }
+}
+```
+
+Finally, add `AsyncDatabendDriver` as a Python class in `lib.rs`:
+
+```rust
+#[pymodule]
+fn _databend_driver(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_class::<AsyncDatabendDriver>()?;
+    Ok(())
+}
+```
+
+If you are interested in learning more, please check out the resources listed below:
+
+- [PR | feat: Support python binding](https://github.com/datafuselabs/bendsql/pull/112)
+
+## Highlights
+
+Here are some noteworthy items recorded here, perhaps you can find something that interests you.
```

Comment:
We have also made these improvements to Databend that we hope you will find helpful: